client/internal/package_download_details_reporter.go

```diff
@@ -0,0 +1,78 @@
+package internal
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	"github.com/open-telemetry/opamp-go/protobufs"
+)
+
+const downloadReporterDefaultInterval = time.Second * 10
+
+type downloadReporter struct {
+	start         time.Time
+	interval      time.Duration
+	packageLength float64
+
+	downloaded atomic.Uint64
+
+	done chan struct{}
+}
+
+func newDownloadReporter(interval time.Duration, length int) *downloadReporter {
+	if interval <= 0 {
+		interval = downloadReporterDefaultInterval
+	}
+	return &downloadReporter{
+		start:         time.Now(),
+		interval:      interval,
+		packageLength: float64(length),
+		done:          make(chan struct{}),
+	}
+}
+
+// Write tracks the number of bytes downloaded. It will never return an error.
+func (p *downloadReporter) Write(b []byte) (int, error) {
+	n := len(b)
+	p.downloaded.Add(uint64(n))
+	return n, nil
+}
+
+// report periodically updates the package status details and calls the passed upateFn to send the new status.
+func (p *downloadReporter) report(ctx context.Context, status *protobufs.PackageStatus, updateFn func(context.Context, bool) error) {
+	go func() {
+		// Make sure we wait at least 1s before reporting the download rate in bps to avoid a panic
+		time.Sleep(time.Second)
+		ticker := time.NewTicker(p.interval)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-p.done:
+				return
+			case <-ticker.C:
+				downloadTime := time.Now().Sub(p.start)
+				downloaded := float64(p.downloaded.Load())
+				bps := downloaded / float64(downloadTime/time.Second)
+				var downloadPercent float64
+				if p.packageLength > 0 {
+					downloadPercent = downloaded / p.packageLength * 100
+				}
+				status.DownloadDetails = &protobufs.PackageDownloadDetails{
```

Comment:
I am a bit worried that `status` is a struct owned by someone else and we modify it from our goroutine here, while `status`'s other fields are also simultaneously being modified by the owner. If the code is accidentally modified in the future to modify the same fields we will have a problem.

Can we instead change `updateFn` to accept the `DownloadPercent` and `DownloadBytesPerSecond` as parameters and eliminate the `status` parameter from `downloadReporter`? This way the responsibility for getting concurrency correctly will be sole responsibility of the caller of `downloadReporter.report` instead of being a shared responsibility.